File: rust/src/sse/claude.rs
use crate::protocol::{ErrorEvent, Event, StopEvent, TextEvent, UsageEvent};
use crate::sse::toolcall::build_tool_call_event;
use anyhow::{Result, anyhow};
use serde_json::Value;
use std::io::{BufRead, BufReader, Read};

#[derive(Default)]
struct SseEvent {
    event: String,
    data: String,
}

fn read_sse<R: Read>(reader: R, mut emit: impl FnMut(SseEvent) -> Result<()>) -> Result<()> {
    let mut br = BufReader::new(reader);
    let mut line = String::new();
    let mut cur = SseEvent::default();
    loop {
        line.clear();
        let n = br.read_line(&mut line)?;
        if n == 0 {
            if !cur.event.is_empty() || !cur.data.is_empty() {
                emit(cur)?;
            }
            break;
        }
        let l = line.trim_end_matches(['\n', '\r']);
        if l.is_empty() {
            if !cur.event.is_empty() || !cur.data.is_empty() {
                emit(cur)?;
            }
            cur = SseEvent::default();
            continue;
        }
        if let Some(v) = l.strip_prefix("event: ") {
            cur.event = v.to_string();
        } else if let Some(v) = l.strip_prefix("data: ") {
            cur.data.push_str(v);
        }
    }
    Ok(())
}

pub fn parse<R: Read>(reader: R, mut emit: impl FnMut(Event) -> Result<()>) -> Result<()> {
    let mut block_type = String::new();
    let mut tool_name = String::new();
    let mut tool_id = String::new();
    let mut partial_json = String::new();
    let mut stop_reason = String::new();
    let mut input_tokens = 0i64;
    let mut output_tokens = 0i64;
    let mut cache_input_tokens = 0i64;

    read_sse(reader, |evt| {
        if evt.data.is_empty() {
            return Ok(());
        }
        let body: Value = serde_json::from_str(&evt.data)?;
        match evt.event.as_str() {
            "content_block_start" => {
                let cb = body.get("content_block").cloned().unwrap_or(Value::Null);
                let t = cb.get("type").and_then(Value::as_str).unwrap_or("");
                if t == "text" {
                    block_type = "text".to_string();
                } else if t == "tool_use" {
                    block_type = "tool".to_string();
                    tool_name = cb
                        .get("name")
                        .and_then(Value::as_str)
                        .unwrap_or("")
                        .to_string();
                    tool_id = cb
                        .get("id")
                        .and_then(Value::as_str)
                        .unwrap_or("")
                        .to_string();
                    partial_json.clear();
                }
            }
            "content_block_delta" => {
                let delta = body.get("delta").cloned().unwrap_or(Value::Null);
                if block_type == "text" {
                    let text = delta.get("text").and_then(Value::as_str).unwrap_or("");
                    if !text.is_empty() {
                        emit(Event::Text(TextEvent {
                            content: text.to_string(),
                        }))?;
                    }
                } else if block_type == "tool" {
                    partial_json.push_str(
                        delta
                            .get("partial_json")
                            .and_then(Value::as_str)
                            .unwrap_or(""),
                    );
                }
            }
            "content_block_stop" => {
                if block_type == "tool" {
                    let call = build_tool_call_event(&tool_name, &tool_id, &partial_json)?;
                    emit(Event::ToolCall(call))?;
                }
                block_type.clear();
            }
            "message_delta" => {
                if let Some(r) = body
                    .get("delta")
                    .and_then(|v| v.get("stop_reason"))
                    .and_then(Value::as_str)
                {
                    if !r.is_empty() {
                        stop_reason = r.to_string();
                    }
                }
                let usage = body.get("usage").cloned().unwrap_or(Value::Null);
                if let Some(v) = usage.get("input_tokens").and_then(Value::as_i64) {
                    input_tokens = v;
                }
                if let Some(v) = usage.get("output_tokens").and_then(Value::as_i64) {
                    output_tokens = v;
                }
                if let Some(v) = usage
                    .get("cache_read_input_tokens")
                    .and_then(Value::as_i64)
                    .or_else(|| {
                        usage
                            .get("cache_creation_input_tokens")
                            .and_then(Value::as_i64)
                    })
                {
                    cache_input_tokens = v;
                }
            }
            "message_start" => {
                let usage = body
                    .get("message")
                    .and_then(|m| m.get("usage"))
                    .cloned()
                    .unwrap_or(Value::Null);
                if let Some(v) = usage.get("input_tokens").and_then(Value::as_i64) {
                    input_tokens = v;
                }
                if let Some(v) = usage
                    .get("cache_read_input_tokens")
                    .and_then(Value::as_i64)
                    .or_else(|| {
                        usage
                            .get("cache_creation_input_tokens")
                            .and_then(Value::as_i64)
                    })
                {
                    cache_input_tokens = v;
                }
            }
            "message_stop" => {
                emit(Event::Usage(UsageEvent {
                    input_tokens,
                    output_tokens,
                    cache_input_tokens,
                }))?;
                emit(Event::Stop(StopEvent {
                    reason: stop_reason.clone(),
                }))?;
            }
            "error" => {
                let msg = body
                    .get("message")
                    .and_then(Value::as_str)
                    .or_else(|| {
                        body.get("error")
                            .and_then(|e| e.get("message"))
                            .and_then(Value::as_str)
                    })
                    .unwrap_or("unknown error");
                emit(Event::Error(ErrorEvent {
                    message: msg.to_string(),
                }))?;
            }
            _ => {}
        }
        Ok(())
    })
    .map_err(|e| anyhow!("parse claude sse: {e}"))?;

    Ok(())
}

File: go/internal/sse/claude.go
package sse

import (
	"bufio"
	"encoding/json"
	"fmt"
	"io"
	"strings"

	"github.com/lloydzhou/bash-agent/internal/protocol"
)

type SSEEvent struct {
	Event string
	Data  string
}

func ReadSSE(r io.Reader, fn func(SSEEvent) error) error {
	scanner := bufio.NewScanner(r)
	buf := make([]byte, 0, 64*1024)
	scanner.Buffer(buf, 1024*1024)
	var current SSEEvent
	for scanner.Scan() {
		line := scanner.Text()
		if line == "" {
			if current.Data != "" || current.Event != "" {
				if err := fn(current); err != nil {
					return err
				}
			}
			current = SSEEvent{}
			continue
		}
		switch {
		case strings.HasPrefix(line, "event: "):
			current.Event = strings.TrimPrefix(line, "event: ")
		case strings.HasPrefix(line, "data: "):
			current.Data += strings.TrimPrefix(line, "data: ")
		}
	}
	if current.Data != "" || current.Event != "" {
		if err := fn(current); err != nil {
			return err
		}
	}
	return scanner.Err()
}

type ClaudeParser struct{}

func (p ClaudeParser) Parse(r io.Reader, emit func(protocol.Event) error) error {
	var (
		blockType        string
		toolName         string
		toolID           string
		partialJSON      string
		stopReason       string
		inputTokens      int
		outputTokens     int
		cacheInputTokens int
	)
	return ReadSSE(r, func(evt SSEEvent) error {
		if evt.Data == "" {
			return nil
		}
		var body map[string]json.RawMessage
		if err := json.Unmarshal([]byte(evt.Data), &body); err != nil {
			return fmt.Errorf("parse claude sse: %w", err)
		}
		switch evt.Event {
		case "content_block_start":
			var payload struct {
				ContentBlock struct {
					Type string `json:"type"`
					Name string `json:"name"`
					ID   string `json:"id"`
				} `json:"content_block"`
			}
			if err := json.Unmarshal([]byte(evt.Data), &payload); err != nil {
				return err
			}
			switch payload.ContentBlock.Type {
			case "text":
				blockType = "text"
			case "tool_use":
				blockType = "tool"
				toolName = payload.ContentBlock.Name
				toolID = payload.ContentBlock.ID
				partialJSON = ""
			}
		case "content_block_delta":
			if blockType == "text" {
				var payload struct {
					Delta struct {
						Text string `json:"text"`
					} `json:"delta"`
				}
				if err := json.Unmarshal([]byte(evt.Data), &payload); err != nil {
					return err
				}
				if payload.Delta.Text != "" {
					if err := emit(protocol.TextEvent{Content: payload.Delta.Text}); err != nil {
						return err
					}
				}
			} else if blockType == "tool" {
				var payload struct {
					Delta struct {
						PartialJSON string `json:"partial_json"`
					} `json:"delta"`
				}
				if err := json.Unmarshal([]byte(evt.Data), &payload); err != nil {
					return err
				}
				partialJSON += payload.Delta.PartialJSON
			}
		case "content_block_stop":
			if blockType == "tool" {
				call, err := BuildToolCallEvent(toolName, toolID, partialJSON)
				if err != nil {
					return err
				}
				if err := emit(call); err != nil {
					return err
				}
			}
			blockType = ""
		case "message_delta":
			var payload struct {
				Delta struct {
					StopReason string `json:"stop_reason"`
				} `json:"delta"`
				Usage struct {
					InputTokens              int `json:"input_tokens"`
					OutputTokens             int `json:"output_tokens"`
					CacheReadInputTokens     int `json:"cache_read_input_tokens"`
					CacheCreationInputTokens int `json:"cache_creation_input_tokens"`
				} `json:"usage"`
			}
			if err := json.Unmarshal([]byte(evt.Data), &payload); err != nil {
				return err
			}
			if payload.Delta.StopReason != "" {
				stopReason = payload.Delta.StopReason
			}
			if payload.Usage.InputTokens != 0 {
				inputTokens = payload.Usage.InputTokens
			}
			if payload.Usage.OutputTokens != 0 {
				outputTokens = payload.Usage.OutputTokens
			}
			if payload.Usage.CacheReadInputTokens != 0 {
				cacheInputTokens = payload.Usage.CacheReadInputTokens
			} else if payload.Usage.CacheCreationInputTokens != 0 {
				cacheInputTokens = payload.Usage.CacheCreationInputTokens
			}
		case "message_start":
			var payload struct {
				Message struct {
					Usage struct {
						InputTokens              int `json:"input_tokens"`
						CacheReadInputTokens     int `json:"cache_read_input_tokens"`
						CacheCreationInputTokens int `json:"cache_creation_input_tokens"`
					} `json:"usage"`
				} `json:"message"`
			}
			if err := json.Unmarshal([]byte(evt.Data), &payload); err != nil {
				return err
			}
			if payload.Message.Usage.InputTokens != 0 {
				inputTokens = payload.Message.Usage.InputTokens
			}
			if payload.Message.Usage.CacheReadInputTokens != 0 {
				cacheInputTokens = payload.Message.Usage.CacheReadInputTokens
			} else if payload.Message.Usage.CacheCreationInputTokens != 0 {
				cacheInputTokens = payload.Message.Usage.CacheCreationInputTokens
			}
		case "message_stop":
			if err := emit(protocol.UsageEvent{
				InputTokens:      inputTokens,
				OutputTokens:     outputTokens,
				CacheInputTokens: cacheInputTokens,
			}); err != nil {
				return err
			}
			return emit(protocol.StopEvent{Reason: stopReason})
		case "error":
			var payload struct {
				Error struct {
					Message string `json:"message"`
				} `json:"error"`
				Message string `json:"message"`
			}
			if err := json.Unmarshal([]byte(evt.Data), &payload); err != nil {
				return err
			}
			msg := payload.Message
			if msg == "" {
				msg = payload.Error.Message
			}
			return emit(protocol.ErrorEvent{Message: msg})
		}
		_ = body
		return nil
	})
}
